File: service/src/whf/eval/harness.py
"""Two-level evaluation: arrival accuracy per model, and demand accuracy of the whole pipeline per model."""

from __future__ import annotations

import datetime as dt
import sqlite3
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path

import numpy as np
import pandas as pd

from whf.backtest import default_origins, interval_bounds, rolling_backtest
from whf.calendar import ONE_WEEK, last_complete_week
from whf.eval.metrics import coverage, weighted_quantile_loss
from whf.eval.truth import realised_hours, truncated_copy, truth_from_answer_key
from whf.models import MODEL_FACTORIES
from whf.models.base import ArrivalModel, ModelUnavailable
from whf.pipeline import MIN_WEEKS_BEFORE_ORIGIN, _load_frames, arrival_feature_matrix, run_forecast

HORIZONS = (1, 2)
METRIC_COLUMNS = ["model", "horizon", "origin", "metric", "value"]
DEMAND_COLUMNS = [
    "model",
    "origin",
    "team_id",
    "member_id",
    "week_start",
    "forecast",
    "truth",
    "capacity",
    "open_hours",
    "new_hours",
]


@dataclass
class EvalConfig:
    as_of: dt.date
    origins: int = 6
    models: tuple[str, ...] = ()
    teams: tuple[int, ...] = ()
    finetune: bool = False
    answer_key: Path | None = None


@dataclass
class EvalResult:
    scores: pd.DataFrame
    demand: pd.DataFrame
    skipped: dict[str, str] = field(default_factory=dict)
    truth_source: str = ""
    elapsed_seconds: float = 0.0
    origins: list[dt.date] = field(default_factory=list)


def _leave_one_out_band(residuals: pd.DataFrame, origin: dt.date) -> tuple[float, float]:
    others = residuals[residuals["origin"] != origin]["residual"].to_numpy(dtype=float)
    return interval_bounds(others) if len(others) else (0.0, 0.0)


def arrival_level(
    feat: pd.DataFrame,
    factories: dict[str, Callable[[], ArrivalModel]],
    origins: list[dt.date],
    horizons: tuple[int, ...],
) -> tuple[pd.DataFrame, dict[str, str]]:
    result = rolling_backtest(feat, factories, origins, horizons)
    rows: list[dict] = []
    per_origin_residuals = result.residual_frames
    for score in result.scores.itertuples(index=False):
        base = {"model": score.model, "horizon": score.horizon, "origin": score.origin}
        rows.append({**base, "metric": "mae", "value": score.mae})
        rows.append({**base, "metric": "mase", "value": score.mase})
        rows.append(
            {
                **base,
                "metric": "beats_naive",
                "value": float(score.mase < 1.0) if not np.isnan(score.mase) else float("nan"),
            }
        )
        key = (score.model, score.horizon)
        res = per_origin_residuals[key]
        mine = res[res.origin == score.origin]["residual"].to_numpy(dtype=float)
        if key in result.quantiles:
            q = result.quantiles[key]
            q = q[q.origin == score.origin]
            y, low, high = (q[c].to_numpy(dtype=float) for c in ("y", "low", "high"))
            point = y - mine
            cov, wql = coverage(y, low, high), weighted_quantile_loss(y, {0.1: low, 0.5: point, 0.9: high})
        else:
            low_off, high_off = _leave_one_out_band(res, score.origin)
            cov = coverage(mine, np.full_like(mine, low_off), np.full_like(mine, high_off))
            wql = float("nan")
        rows.append({**base, "metric": "coverage80", "value": cov})
        rows.append({**base, "metric": "wql", "value": wql})
        rows.append(
            {**base, "metric": "seconds", "value": result.timings.get(score.model, float("nan")) / max(len(origins), 1)}
        )
    return pd.DataFrame(rows, columns=METRIC_COLUMNS), dict(result.unavailable)


def demand_level(
    conn: sqlite3.Connection,
    factories: dict[str, Callable[[], ArrivalModel]],
    origins: list[dt.date],
    teams: tuple[int, ...],
    truth: pd.DataFrame,
) -> tuple[pd.DataFrame, dict[str, str]]:
    skipped: dict[str, str] = {}
    rows: list[dict] = []
    truth_index = {
        (int(m), w): float(h) for m, w, h in zip(truth.member_id, truth.week_start, truth.hours, strict=True)
    }
    for origin in origins:
        as_of = origin + ONE_WEEK
        replay = truncated_copy(conn, as_of)
        team_ids = teams or tuple(int(t) for t in _load_frames(replay)["teams"]["id"])
        for name in factories:
            if name in skipped:
                continue
            for team_id in team_ids:
                try:
                    result = run_forecast(replay, team_id=team_id, as_of=as_of, force_model=name, persist=False)
                except ModelUnavailable as exc:
                    skipped[name] = str(exc)
                    break
                except ValueError:
                    continue  # team without counted members
                for r in result.forecasts.itertuples(index=False):
                    rows.append(
                        {
                            "model": name,
                            "origin": origin,
                            "team_id": team_id,
                            "member_id": int(r.member_id),
                            "week_start": r.week_start,
                            "forecast": float(r.demand_hours),
                            "truth": truth_index.get((int(r.member_id), r.week_start), 0.0),
                            "capacity": float(r.capacity_hours),
                            "open_hours": float(r.open_task_hours),
                            "new_hours": float(r.new_task_hours),
                        }
                    )
    return pd.DataFrame(rows, columns=DEMAND_COLUMNS), skipped


def evaluate(
    conn: sqlite3.Connection,
    config: EvalConfig,
    factories: dict[str, Callable[[], ArrivalModel]] | None = None,
) -> EvalResult:
    started = time.perf_counter()
    factories = factories or MODEL_FACTORIES
    for name in config.models:
        if name not in factories:
            raise ValueError(f"unknown model {name!r}; known: {sorted(factories)}")
    chosen = {name: factories[name] for name in (config.models or tuple(factories))}
    if config.finetune:
        # Harness-only candidate: the same adapter, fine-tuned on the training window of every origin.
        from whf.models.chronos2 import Chronos2Arrival

        chosen["chronos2_ft"] = lambda: Chronos2Arrival(finetune=True)
    frames = _load_frames(conn)
    origin = last_complete_week(config.as_of)
    _, feat, weeks = arrival_feature_matrix(frames, origin)
    origins = [o for o in default_origins(origin, config.origins) if o >= weeks[0] + MIN_WEEKS_BEFORE_ORIGIN * ONE_WEEK]
    scores, skipped_a = arrival_level(feat, chosen, origins, HORIZONS)
    if config.answer_key is not None:
        truth, source = truth_from_answer_key(config.answer_key), "answer key"
    else:
        truth, source = realised_hours(conn), "realised hours"
    demand, skipped_b = demand_level(
        conn, {n: f for n, f in chosen.items() if n not in skipped_a}, origins, config.teams, truth
    )
    return EvalResult(
        scores=scores,
        demand=demand,
        skipped={**skipped_a, **skipped_b},
        truth_source=source,
        elapsed_seconds=time.perf_counter() - started,
        origins=origins,
    )

File: server/forecast-core/src/main/java/com/workloadhub/forecast/eval/Harness.java
package com.workloadhub.forecast.eval;

import com.workloadhub.forecast.api.ForecastException;
import com.workloadhub.forecast.api.MemberWindowForecast;
import com.workloadhub.forecast.backtest.Backtest;
import com.workloadhub.forecast.calendar.ForecastWindow;
import com.workloadhub.forecast.calendar.Weeks;
import com.workloadhub.forecast.calendar.WorkingCalendar;
import com.workloadhub.forecast.capacity.CapacityRule;
import com.workloadhub.forecast.data.ForecastData;
import com.workloadhub.forecast.data.rows.TeamRow;
import com.workloadhub.forecast.features.FeatureBuilder;
import com.workloadhub.forecast.features.FeatureMatrix;
import com.workloadhub.forecast.features.MemberDay;
import com.workloadhub.forecast.features.MemberWeek;
import com.workloadhub.forecast.lifecycle.Lifecycle;
import com.workloadhub.forecast.lifecycle.Truncation;
import com.workloadhub.forecast.run.ForecastRunner;
import com.workloadhub.forecast.run.Prepared;
import com.workloadhub.forecast.run.TeamOutcome;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.SortedMap;
import java.util.UUID;
import java.util.stream.IntStream;

/**
 * Two-level evaluation: arrival accuracy of the single booster, and demand accuracy of the whole pipeline.
 *
 * <p>The model tournament this class used to run is gone with the tournament itself (design 2026-09-13): there is
 * one model, so no row this class produces carries a model column any more, and a caller cannot ask to evaluate one
 * model over another — {@code ModelUnavailable} (the native library missing, or too few training rows) is no
 * longer caught and reported as "skipped"; it propagates, because there is nothing left to fall back to
 * (design section 19.6, an accepted cost).
 */
public final class Harness {

    private final ForecastRunner runner;
    private final CapacityRule rule;

    public Harness(ForecastRunner runner, CapacityRule rule) {
        this.runner = runner;
        this.rule = rule;
    }

    public EvalResult evaluate(ForecastData data, EvalConfig configIn) {
        long started = System.nanoTime();
        EvalConfig config = configIn.asOf() != null ? configIn
                : new EvalConfig(lastCreated(data).orElseGet(LocalDate::now), configIn.origins(), configIn.teams(), configIn.windows());
        // Both levels must measure the same forecast: Level A takes its horizons from config.windows() while
        // Level B is the injected runner, which computes every run at its own whf.forecast.windows. A caller
        // asking for a count the runner does not have would get a report whose header says one number and whose
        // demand rows stop at another; the driver escapes by construction, a host through the public API does not.
        if (config.windows() != runner.windows()) {
            throw ForecastException.of("INVALID_REQUEST", "evaluation asks for " + config.windows()
                    + " windows but the runner forecasts " + runner.windows());
        }
        LocalDate origin = Weeks.lastCompleteWeek(config.asOf());
        Lifecycle lc = Lifecycle.derive(data);
        WorkingCalendar cal = WorkingCalendar.fromHolidays(data.holidays());
        int windows = config.windows();
        // The horizons this evaluation scores: one per window (1..windows), the shape HORIZONS = {1, 2} always
        // had at the old fixed window count of two. This is narrower than Features.horizons(windows), which pads
        // to windows + 1 for the feature matrix a live run needs; the evaluation harness has no such run-day
        // edge case to cover, and padding it here starves leave-one-out coverage at the extra horizon with few
        // origins.
        int[] horizons = IntStream.rangeClosed(1, windows).toArray();
        int maxHorizon = windows;
        FeatureMatrix features = new FeatureBuilder(data, lc, cal, rule, windows).build(data.members(), origin);
        LocalDate firstWeek = features.keys().stream().map(MemberWeek::week).min(LocalDate::compareTo).orElse(origin);
        List<LocalDate> origins = features.rowCount() == 0 ? List.of() : Backtest.origins(origin, firstWeek, config.origins(), maxHorizon);
        Backtest.Result bt = Backtest.run(features, origins, horizons);
        List<ScoreRow> scores = arrivalLevel(bt, horizons);
        List<DemandRow> demand = demandLevel(data, origins, config.teams());
        return new EvalResult(scores, demand, Truth.SOURCE, (System.nanoTime() - started) / 1e9, origins, config, fingerprint(data));
    }

    /** What the evaluation ran against, for the report's reproducibility section. */
    static Map<String, String> fingerprint(ForecastData data) {
        Map<String, String> f = new LinkedHashMap<>();
        f.put("members", String.valueOf(data.members().size()));
        f.put("teams", String.valueOf(data.teams().size()));
        f.put("tasks", String.valueOf(data.tasks().size()));
        f.put("time_logs", String.valueOf(data.timeLogs().size()));
        f.put("first_created", data.tasks().stream().map(t -> t.createdDate().toLocalDate()).min(LocalDate::compareTo).map(Object::toString).orElse("none"));
        f.put("last_created", lastCreated(data).map(Object::toString).orElse("none"));
        return f;
    }

    private static Optional<LocalDate> lastCreated(ForecastData data) {
        return data.tasks().stream().map(t -> t.createdDate().toLocalDate()).max(LocalDate::compareTo);
    }

    /** {@code mase} and {@code beats_naive} are gone with the tournament's naive floor (task 5); {@code mae}, the interval metrics and
     * {@code seconds} are what remains scored. */
    static List<ScoreRow> arrivalLevel(Backtest.Result bt, int[] horizons) {
        List<ScoreRow> rows = new ArrayList<>();
        for (Backtest.Score s : bt.scores()) {
            rows.add(new ScoreRow(s.horizon(), s.origin(), "mae", s.mae()));
            List<Backtest.Residual> all = bt.residualRows(s.horizon());
            double[] others = all.stream().filter(r -> !r.origin().equals(s.origin())).mapToDouble(Backtest.Residual::residual).toArray();
            List<Backtest.Residual> mine = all.stream().filter(r -> r.origin().equals(s.origin())).toList();
            double coverage = Double.NaN;
            double wql = Double.NaN;
            if (others.length > 0 && !mine.isEmpty()) {
                double[] q = Backtest.intervalBounds(others);
                double[] y = mine.stream().mapToDouble(Backtest.Residual::y).toArray();
                double[] point = mine.stream().mapToDouble(r -> r.y() - r.residual()).toArray();
                double[] low = new double[y.length];
                double[] high = new double[y.length];
                for (int i = 0; i < y.length; i++) {
                    low[i] = point[i] + q[0];
                    high[i] = point[i] + q[1];
                }
                coverage = Metrics.coverage(y, low, high);
                wql = Metrics.weightedQuantileLoss(y, Map.of(0.1, low, 0.5, point, 0.9, high));
            }
            rows.add(new ScoreRow(s.horizon(), s.origin(), "coverage80", coverage));
            rows.add(new ScoreRow(s.horizon(), s.origin(), "wql", wql));
            rows.add(new ScoreRow(s.horizon(), s.origin(), "seconds", s.horizon() == horizons[0] ? bt.seconds() / Math.max(1, bt.scores().size() / horizons.length) : Double.NaN));
        }
        return rows;
    }

    private List<DemandRow> demandLevel(ForecastData data, List<LocalDate> origins, List<UUID> teamsIn) {
        SortedMap<MemberDay, Double> truth = Truth.realisedHoursByDay(data);
        List<DemandRow> rows = new ArrayList<>();
        List<UUID> teams = teamsIn.isEmpty()
                ? data.teams().stream().map(TeamRow::id).filter(t -> !data.membersOfTeam(t).isEmpty()).sorted((a, b) -> a.toString().compareTo(b.toString())).toList()
                : teamsIn;
        for (LocalDate origin : origins) {
            LocalDate asOf = origin.plusWeeks(1);
            ForecastData replay = Truncation.at(data, asOf);
            Prepared prepared = runner.prepare(replay, asOf, ForecastRunner.ProgressListener.NONE);
            for (UUID team : teams) {
                TeamOutcome outcome;
                try {
                    outcome = runner.forTeam(prepared, team);
                } catch (ForecastException e) {
                    if ("TEAM_NOT_FOUND".equals(e.code())) {
                        continue;
                    }
                    throw e;
                }
                for (MemberWindowForecast w : outcome.memberWindows()) {
                    ForecastWindow window = prepared.windows().get(w.windowIndex() - 1);
                    double realised = 0;
                    for (LocalDate d : window.weekdays()) {
                        realised += truth.getOrDefault(new MemberDay(w.userId(), d), 0.0);
                    }
                    rows.add(new DemandRow(origin, team, w.userId(), w.windowIndex(), w.windowStart(), w.windowEnd(), w.demandHrs(),
                            Math.round(realised * 1e6) / 1e6, w.capacityHrs(), w.backlogExcessHrs(), w.dueExcessHrs()));
                }
            }
        }
        return rows;
    }
}
